### Backtest Statistics

Most of the functions below can be found under:

* Backtest/stats_measure
* Sample_data/make_data

A quick slideshow: [Seven Sins of Backtesting](https://newyork.qwafafew.org/wp-content/uploads/sites/4/2015/10/Luo_20150128.pdf)

If you are keen on generating synthetic data for your research, copy the code snippets [Generate synthetic raw data](https://gist.github.com/boyboi86/5e00faf48f60abfdbe838fbdee269471) in my gist.

[Additional resource](https://en.wikipedia.org/wiki/Euler%E2%80%93Mascheroni_constant)

Contact: boyboi86@gmail.com

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import research as rs

%matplotlib inline

Num of CPU core:  4
Machine info:  Windows-10-10.0.18362-SP0
Python 3.7.4 (default, Aug  9 2019, 18:34:13) [MSC v.1915 64 bit (AMD64)]
Numpy 1.18.5
Pandas 1.0.4


In [2]:
portfolio_mtx = rs.create_portfolio(price = [95, 1000], 
                                 position = [1000, 10000], 
                                 n_sample = 2000, 
                                 imbalance = True)

Sample Portfolio Construct:
Equity label: 1773
Bond label: 227
Equity to Debt Ratio: 7.8106

Junk bond (Below BBB-grade): 11.35 %
Dividend equity: 88.65 %


**Note**

The below is our entire portfolio universe, expressed in matrix format.

Sythetic portfolio assumptions:
* In this portfolio, there are 10,000 assets (Label 0 = equity, label 1 = bond).
* All long positions, however if you are doing both long/ short style investing. Just change sign of position.
* Guaranteed to have dividend or interest.
* Clean price are meant for bond assets, for equity just refer to open/ close price.
* Formula for bond yield:

        interest = dirty close - clean close
        
* Position size will not turn from negative to positive and vice-versa (No flipping).
* Open price can be replaced with average transacted price, if you were executing average cost pricing.
* If average cost pricing method is employed, only account for single direction purchase.
* Each position can only reflect 1 direction.
* index are beginning period, t1 is the ending period.
* The below are expressed in single unit of asset:
    1. Open
    2. Close
    3. Yield
    4. Clean Close
    5. Clean Open


* Junk bonds and dividend equities are expressed in percentage of entire portfolio.
* Clean and dirty price is unique to only bonds (Label = 1).
* Yield is expressed as a nominal term (Based on close price).
* If profit/ loss made in sale/ purchase from long/ short position will be reflected in yield.
* All yield must be positive, the function will reflect short sales.
* Expense represents all costs incurred till date that is due to asset.
* Expense is expressed in nominal terms, which consist:

    1. Taxes
    2. Transaction fees
    3. Out-of-pockets
    4. Custodian charges
    5. Margin Facilities
    6. Any charges/ expenses
    
    
**Note**

If you are shorting instead, it will not increase your AUM but it will increase profit/ loss.

**Note**

This is not from the book. But I feel that it's necessary for me since I do not have S&P 500, to finish this tutorial.

In [3]:
portfolio_mtx

t1        open       close  \
2012-10-18 16:48:42.537666 2012-10-23 20:36:38.274374  712.341641  591.554442   
2012-10-19 16:48:42.537666 2012-10-26 12:09:08.160024  237.697665  198.375014   
2012-10-22 16:48:42.537666 2012-10-30 18:47:16.978825  337.199346  384.569274   
2012-10-23 16:48:42.537666 2012-10-30 23:42:13.197475  498.166803  569.613622   
2012-10-24 16:48:42.537666 2012-11-03 00:21:43.647673  377.400237  328.747226   
...                                               ...         ...         ...   
2020-06-11 16:48:42.537666 2020-06-12 23:06:59.648676  123.455649  136.305746   
2020-06-12 16:48:42.537666 2020-06-14 07:59:29.097476  451.172391  427.303924   
2020-06-15 16:48:42.537666 2020-06-21 11:50:42.088945  779.196802  898.547842   
2020-06-16 16:48:42.537666 2020-06-26 13:02:44.560157  814.751467  812.444718   
2020-06-17 16:48:42.537666 2020-06-26 06:45:14.294269  587.205694  559.298560   

                            open_pos  close_pos      yield    expense  label  \
2012-10-18 16:48:42.537666    4914.0     3205.0  39.150070  33.009637    0.0   
2012-10-19 16:48:42.537666    9161.0     6750.0  10.015237   6.405694    0.0   
2012-10-22 16:48:42.537666    1381.0     3611.0  26.156680  13.082456    0.0   
2012-10-23 16:48:42.537666    6083.0     9891.0  21.248672   4.134028    0.0   
2012-10-24 16:48:42.537666    4013.0     2955.0  17.097073  18.412937    0.0   
...                              ...        ...        ...        ...    ...   
2020-06-11 16:48:42.537666    5323.0     5670.0   6.113479   2.186776    1.0   
2020-06-12 16:48:42.537666    6690.0     9007.0  49.815237   2.481167    0.0   
2020-06-15 16:48:42.537666    4260.0     3949.0  54.633637  27.484208    0.0   
2020-06-16 16:48:42.537666    9597.0     5006.0  27.454042  24.794070    0.0   
2020-06-17 16:48:42.537666    4923.0     4932.0  46.968234   8.365209    0.0   

                            clean_open  clean_close  
2012-10-18 16:48:42.537666  664.491780   552.404372  
2012-10-19 16:48:42.537666  224.334496   188.359777  
2012-10-22 16:48:42.537666  319.682111   358.412593  
2012-10-23 16:48:42.537666  459.372735   548.364950  
2012-10-24 16:48:42.537666  356.027994   311.650153  
...                                ...          ...  
2020-06-11 16:48:42.537666  116.760033   130.192266  
2020-06-12 16:48:42.537666  415.737965   377.488687  
2020-06-15 16:48:42.537666  732.210613   843.914205  
2020-06-16 16:48:42.537666  762.422269   784.990675  
2020-06-17 16:48:42.537666  553.458598   512.330327  

[2000 rows x 10 columns]

In [4]:
def rtn_by_tw(mtx: pd.DataFrame):
    df0 = pd.DataFrame(index=mtx.index).assign(t1 = mtx.t1)
    mtx['equity_diff'] = mtx['close'].sub(mtx['open'])
    mtx['bond_diff'] = mtx['clean_close'].sub(mtx['clean_open'])
    for idx in mtx.index:
        _close_pos = mtx.loc[idx, 'close_pos']
        _open_pos = mtx.loc[idx, 'open_pos']
        if ((_open_pos) > .0 and (_close_pos) >= .0) or ((_open_pos) < .0 and (_close_pos) >= .0):
            mtx.loc[idx, 't1_pos'] = max(0, float(_close_pos - _open_pos))
            mtx.loc[idx, 't1_vol'] = float(_open_pos - _close_pos)
        if ((_open_pos) > .0 and (_close_pos) <= .0) or ((_open_pos) > .0 and (_close_pos) <= .0):
            mtx.loc[idx, 't1_pos'] = max(0, float(_close_pos - _open_pos)) #shorting doesn't increase your aum
            mtx.loc[idx, 't1_vol'] = float(_open_pos - _close_pos)
            mtx[idx, 'yield'].mul(-1) #if you short; you will pay for the div/ interest instead
        
        #but it does add to your profit.. or losses
        if mtx.loc[idx, 'label'] == 0:
            mtx.loc[idx, 'pnl'] = (mtx.loc[idx, 'bond_diff'] + mtx.loc[idx, 'yield']) * mtx.loc[idx, 't1_vol']
        else:
            mtx.loc[idx, 'pnl'] = (mtx.loc[idx, 'equity_diff'] + mtx.loc[idx, 'yield']) * mtx.loc[idx, 't1_vol']
            
    
    t0_val = mtx['close'].mul(mtx['open_pos'])
    t1_val = mtx['close'].mul(mtx['t1_pos'])
    
    df0['rtn'] = mtx['pnl'].sub(mtx['expense']).div(t0_val.add(t1_val))
    df0['rtn'].add(1)

    return df0

In [5]:
df2 = rtn_by_tw(mtx = portfolio_mtx)

df2

t1       rtn
2012-10-18 16:48:42.537666 2012-10-23 20:36:38.274374 -0.042892
2012-10-19 16:48:42.537666 2012-10-26 12:09:08.160024 -0.034444
2012-10-22 16:48:42.537666 2012-10-30 18:47:16.978825 -0.104208
2012-10-23 16:48:42.537666 2012-10-30 23:42:13.197475 -0.074512
2012-10-24 16:48:42.537666 2012-11-03 00:21:43.647673 -0.021892
...                                               ...       ...
2020-06-11 16:48:42.537666 2020-06-12 23:06:59.648676 -0.008517
2020-06-12 16:48:42.537666 2020-06-14 07:59:29.097476 -0.006964
2020-06-15 16:48:42.537666 2020-06-21 11:50:42.088945  0.013507
2020-06-16 16:48:42.537666 2020-06-26 13:02:44.560157  0.029451
2020-06-17 16:48:42.537666 2020-06-26 06:45:14.294269 -0.000022

[2000 rows x 2 columns]

**Note**

Based on the above, we will have to run the next function using parallelization, so that we can link these performance geometrically and weight these returns based on time.

Otherwise, it may take much longer than expected.

If you are interested in parallelization but want some demostration.

[AFML 13.1](https://github.com/boyboi86/AFML/blob/master/AFML%2013.1.ipynb)

**Note**

Running parallel operations for binary integer is significantly faster than float.

If required try downcast dtype (Last resort).

In [6]:
# After parallelization!

tw_rtn = rs.rtn_by_tw(mtx = portfolio_mtx, num_threads=3)

[                            0     1     2     3     4     5     6     7     \
2015-05-13 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 10:02:26.269215   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 17:20:26.760815   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 21:08:17.860822   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
...                          ...   ...   ...   ...   ...   ...   ...   ...   
2017-11-27 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-28 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-29 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-30 13:06:22.281616   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-30 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   

                            8     9     ...  1990  1991  1992 

[                            0     1     2     3     4     5     6     7     \
2015-05-13 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 10:02:26.269215   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 17:20:26.760815   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2015-05-14 21:08:17.860822   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
...                          ...   ...   ...   ...   ...   ...   ...   ...   
2017-11-27 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-28 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-29 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-30 13:06:22.281616   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   
2017-11-30 16:48:42.537666   1.0   1.0   1.0   1.0   1.0   1.0   1.0   1.0   

                            8     9     ...  1990  1991  1992 

2020-06-17 16:50:00.111380 100.0% _rtn_by_tw done after 1.18 mins. Remaining 0.0 mins.


**Note**

Assuming all the asset and transaction was provided by the portfolio manager.

The below is the final outcome for the portfolio performance over her career.

If the return is negative at any time, it means the manager is having a negative return during that period.

In [7]:
tw_rtn

2012-10-18 16:48:42.537666   -0.042892
2012-10-19 16:48:42.537666   -0.038677
2012-10-22 16:48:42.537666   -0.061037
2012-10-23 16:48:42.537666   -0.064424
2012-10-23 20:36:38.274374   -0.064424
                                ...   
2020-06-16 16:48:42.537666    0.021448
2020-06-17 16:48:42.537666    0.014241
2020-06-21 11:50:42.088945    0.014241
2020-06-26 06:45:14.294269    0.014607
2020-06-26 13:02:44.560157    0.029451
Length: 4000, dtype: float64

**Note**

Key high-frequency trading strategy charactertistics:
1. low return on transaction cost
2. short holding period
3. High number of bets

If we factor high leverage as part of cost (Margin facilities), each transaction will bear additional cost.

This will increase as part of it's variable cost per transaction basis, but if strategy also exhibits:

1. High turnover
2. High sharpe ratio

There is a good chance that it will have a large capacity (High profitability) even with leverage, in fact leverage can further increase it's profit ratios.

In [8]:
df1 = pd.DataFrame({'close': tw_rtn}).resample('M').mean().fillna(0)

df1

#we have about 93 months of data

close
2012-10-31 -0.037441
2012-11-30  0.007150
2012-12-31 -0.005919
2013-01-31 -0.006310
2013-02-28  0.007525
...              ...
2020-02-29  0.012841
2020-03-31  0.005994
2020-04-30 -0.008560
2020-05-31 -0.018319
2020-06-30  0.009112

[93 rows x 1 columns]

In [9]:
positive_hhi, negative_hhi, time_hhi = rs.hh_idx(rtn = df1, freq = 'M')

sr = rs.sharpe_ratio(rtn = df1.squeeze(), rf_param = .0, t_days = 252)
ir = rs.inform_ratio(rtn = df1.squeeze(), benchmark = .0, t_days = 252)

print("\nSharpe Ratio: {0:.6f}\nInformation Ratio: {1:.6f}".format(sr, ir))

# Almost quite guarantee this portfolio manager will lose your money


Herfindahl-Hirschman Index (M frequency)
HHI positive: 0.031088
HHI Negative: 0.016851
HHI Time between: 0.000000


Sharpe Ratio: -1.798303
Information Ratio: -1.798303


In [10]:
from scipy.stats import skew, kurtosis

pct = rs.proba_sr(obs_sr = sr,
             benchmark_sr = -1.9, # since the sr was negative to begin with
             num_returns = 93, #monthly basis
             skew_returns = df1.skew(),
             kurt_returns = df1.kurtosis())

print("The probability indicates that our sharpe ratio: {0:.6f} is not correct\nIt is {1} which is much lesser than 0.95".format(sr, pct[0]))

The probability indicates that our sharpe ratio: -1.798303 is not correct
It is 0.8261832625852505 which is much lesser than 0.95


In [11]:
pct = rs.proba_sr(obs_sr = sr,
             benchmark_sr = -2.15, # since the sr was negative to begin with
             num_returns = 93, #monthly basis
             skew_returns = df1.skew(),
             kurt_returns = df1.kurtosis())

print("The probability indicates that our sharpe ratio: {0:.6f} is correct\nIt is {1} more than 0.95".format(sr, pct[0]))

The probability indicates that our sharpe ratio: -1.798303 is correct
It is 0.9994188490783524 more than 0.95


**Note**

Notice the change in benchmark can reflect how correct the sharpe ratios are.

In the first example, we set benchmark as 0. We have a score much lesser than 0.95

But when we change the benchmark to be below our existing sharp ratio, it became "correct".

Like-wise if we were to set a benchmark closer to the target sharpe ratio, it becomes more "reliable". (0.9994188490783524)

**Note**

For deflated SR, notice how high is the benchmark as a result deflated sharpe ratio was reduced.

In [12]:
Var_ = 0.5
n_trials = 100

dsr, benchmark = rs.deflated_sr(obs_sr = sr,
                                sr_est = [Var_**(1/2), n_trials],
                                num_returns = 93,
                                skew_returns = df1.skew(), 
                                kurt_returns = df1.kurtosis(),
                                est_param = True)

print("Deflated Sharpe Ratio Probability: {0}\nBenchmark set: {1}".format(dsr[0], benchmark))

Deflated Sharpe Ratio: 4.980385988784981e-241
Benchmark set: 1.7894064662732079


**Note**

In the book, Dr Marcos mentioned time weighted return calculation for a single portfolio. 

Which was what we used previously, however the remaining exercise is more directed at individual securities.

But for completeness, I load whatever S&P500 data I have for the below.

In [13]:
df = pd.read_csv('./research/Sample_data/dollar_bars.txt', 
                 sep=',', 
                 header=0, 
                 parse_dates = True, 
                 index_col=['date_time'])

df = df['close'].to_frame()

draw_dn, time_ddn = rs.drawdn_period(data = df, dollars = True, q_param = .95)

a_rtn = rs.annualized_rtn(data = df.squeeze(), t_days = 252, verbose = True)


Loss Metric (0.95 quartile)
Drawdown: $56.20
Time under water (Based on 252 days): 12.205734


Accuracy Metric Estimate
Total postive returns: 11579
Total negative returns: 12500

Average postive returns: 0.000998
Average negative returns: -0.000916



### Conclusion

**Probabilistic Sharpe Ratio**

Key factor which may influence it's probability outcome are benchmark level. 

If it is too far off from mean (Benchmark), chances are it is not real. (Regardless high or low sharpe ratio)

**Deflated Sharpe Ratio**

If the number of trial were high, it will result in higher benchmark as a result. 

Sharp ratio probability will be very depressed.

**Interesting Fact**

When I tried to run the synthetic data generator to create a portofolio based on random uniform distribution.

I have never gotten once a "portfolio" with performance that has positive performance, the current portfolio in this exercise has the best "performance".

At the very least, it does not return a NaN probability.

Indirectly, it means it is almost impossible to generate alpha on random basis (Luck).